File: how_about_c/raw-nano/src/timer.c
#include "timer.h"
#include<stdio.h>


void timer1_init(void) {
    // turn on timer 1, set prescaler to 64
    TCCR1B = 3; 
}

unsigned short timer1_get_count(void) {
    unsigned short ticks = TCNT1;
    return ticks;
}

void timer1_reset(void) {
    TCNT1 = 0;
}

int timer1_overflow_flag(void) {
    if ( (TIFR1 & (1 << TOV1)) == 0 ) {
        return 0;
    }
    else {
        TIFR1 = (1 << TOV1);
        return 1;
    }
}




void delay(float ms) {
    #warning delay() is both blocking and inefficient

    // this math should not be done on the mcu...
    const float desired_ticks = (uint32_t)(ms*TICKS_PER_MS);
    const uint8_t desired_overflows = (uint8_t)(floor(desired_ticks/MAX_TICKS));
    const uint32_t remaining_ticks = (uint32_t)desired_ticks % MAX_TICKS;
    uint8_t current_overflow = 0;
    unsigned short current_ticks; 
    
    timer1_init();
    while(1) {
        current_ticks = timer1_get_count();
        
        if (desired_overflows > 0) {
            if (timer1_overflow_flag() == 1) {
                if (current_overflow < desired_overflows) {
                    current_overflow += 1;
                    timer1_reset();
                }
                else {
                    if (current_ticks >= remaining_ticks) {
                        break;
                    }
                }
            }
        } 
        else {
            if (current_ticks >= desired_ticks) {
                break;
            }
        }
    }
}

File: C_version/src/timer.rs
use crate::atmega328p::*;
use core::ptr::read_volatile;
use core::ptr::write_volatile;
use micromath::F32Ext;


// For now this is just timer1 which is a 16 bit timer
pub struct Timer {
    pre: u16
}

impl Timer {
   
    pub unsafe fn init(&self) {

        let pre:u8 = match &self.pre{ 
            1 => 1u8,
            8 => 2u8,
            64 => 3u8,
            256 => 4u8,
            1024 => 5u8,
            _ => loop {}
        };

        write_volatile(TCCR1B,pre); // set prescaler to 64 
    }

    pub unsafe fn get_count(&self) -> u16 {
        read_volatile(TCNT1)
    }

    pub unsafe fn reset(&self) {
        write_volatile(TCNT1,0u16);
    }

    pub unsafe fn overflow_flag(&self) -> bool {
        if (read_volatile(TIFR1) & (1 << *TOV1)) == 0 {
            false    
        }
        else {
            write_volatile(TIFR1, 1 << *TOV1);
            true 
        }
        
    }

}

pub const T1: Timer = Timer{pre: 64};
pub const MAX_TICKS: u32 = 65536;
pub const TICKS_PER_MS: u8 = 250;

// #[macro_export]
// macro_rules! delay {
//     ($ms:expr) => {
//         let desired_ticks: u32 = (ms*TICKS_PER_MS) as u32;
//         let desired_overflows: u8 = (((desired_ticks/MAX_TICKS) as f32).floor() ) as u8;
//         let remaining_ticks: u32 =  desired_ticks % MAX_TICKS;
//         let mut current_overflow: u8 = 0; 
//         let mut current_ticks: u32;
        
//         T1.init();
        
//         loop {

//             current_ticks = T1.get_count() as u32;
            
//             if desired_overflows > 0  {
//                 if T1.overflow_flag() == true {
//                     if current_overflow < desired_overflows{
//                         current_overflow += 1;
//                         T1.reset();
//                     }
//                     else {
//                     if current_ticks >= remaining_ticks {
//                             break
//                     }
//                     }
//                 } 
//             }

//             else {
//                 if current_ticks >= desired_ticks {
//                     break
//                 }
//             }
            
//         }

        
//     };
//     () => {
//         loop {};
//     };
// }


pub unsafe fn delay(ms: f32) {
    
    // let desired_ticks = 62500;
    // let desired_overflows = 0;
    // let remaining_ticks = 0;

    let desired_ticks: u32 = (ms*TICKS_PER_MS) as u32;
    let desired_overflows: u8 = (((desired_ticks/MAX_TICKS) as f32).floor() ) as u8;
    let remaining_ticks: u32 =  desired_ticks % MAX_TICKS;
    let mut current_overflow: u8 = 0; 
    let mut current_ticks: u32;

    T1.init();
    
    loop {

        current_ticks = T1.get_count() as u32;
        
        if desired_overflows > 0  {
            if T1.overflow_flag() == true {
                if current_overflow < desired_overflows{
                    current_overflow += 1;
                    T1.reset();
                }
                else {
                if current_ticks >= remaining_ticks {
                        break
                }
                }
            } 
        }

        else {
            if current_ticks >= desired_ticks {
                break
            }
        }
        
    }
}
